External/Plugins/ASCompletion/PluginMain.cs

```diff
@@ -446,104 +446,107 @@ public void HandleEvent(Object sender, NotifyEvent e, HandlingPriority priority)
-                                e.Handled = true;
-                            }
-                            else if (command == "ASCompletion.CtrlAltSpace")
-                            {
-                                ASComplete.OnShortcut(Keys.Control | Keys.Alt | Keys.Space, ASContext.CurSciControl);
-                                e.Handled = true;
-                            }
-                            else if (command == "ASCompletion.ContextualGenerator")
-                            {
-                                if (ASContext.HasContext && ASContext.Context.IsFileValid)
+                                // resolve element under cursor and send a CustomData event
+                                else if (command == "ASCompletion.ResolveElement")
+                                {
+                                    ASComplete.ResolveElement(sci, cmdData);
+                                    e.Handled = true;
+                                }
+                                else if (command == "ASCompletion.MakeIntrinsic")
+                                {
+                                    ASContext.Context.MakeIntrinsic(cmdData);
+                                    e.Handled = true;
+                                }
+
+                                // alternative to default shortcuts
+                                else if (command == "ASCompletion.CtrlSpace")
+                                {
+                                    ASComplete.OnShortcut(Keys.Control | Keys.Space, ASContext.CurSciControl);
+                                    e.Handled = true;
+                                }
+                                else if (command == "ASCompletion.CtrlShiftSpace")
                                 {
-                                    ASGenerator.ContextualGenerator(ASContext.CurSciControl);
+                                    ASComplete.OnShortcut(Keys.Control | Keys.Shift | Keys.Space, ASContext.CurSciControl);
+                                    e.Handled = true;
+                                }
+                                else if (command == "ASCompletion.CtrlAltSpace")
+                                {
+                                    ASComplete.OnShortcut(Keys.Control | Keys.Alt | Keys.Space, ASContext.CurSciControl);
+                                    e.Handled = true;
+                                }
+                                else if (command == "ASCompletion.ContextualGenerator")
+                                {
+                                    if (ASContext.HasContext && ASContext.Context.IsFileValid)
+                                    {
+                                        ASGenerator.ContextualGenerator(ASContext.CurSciControl);
+                                        AddRefactorMenus(ASGenerator.KnownList, de.Data as ToolStripMenuItem);
```

Comment:
Didn't think of that, thank you!
I couldn't really be bothered to look into the event mechanism thinking it would be very complicated, but it seems like it is very similar to AS3 style of event handling with event types as enums and optional data. 
Using `new DataEvent(EventType.Command, "CodeRefactor.AddRefactorOptions", options)` would be alright? I have no idea about the rules for `DataEvent.Action` string.